sumpy/expansion/local.py

```diff
@@ -912,34 +939,67 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
             if self.use_preprocessing_for_m2l:
                 assert m2l_translation_classes_dependent_data is not None
                 assert len(derivatives) == len(src_coeff_exprs)
-                for a, b in zip(derivatives, src_coeff_exprs):
-                    translated_coeffs.append(a * b)
-                return translated_coeffs
-
-            src_coeff_exprs = self.m2l_preprocess_multipole_exprs(src_expansion,
-                    src_coeff_exprs, sac, src_rscale)
-
-            for j in self.get_coefficient_identifiers():
-                translated_coeffs.append(
+                translated_coeffs = [a * b for a, b in zip(derivatives,
+                    src_coeff_exprs)]
+            else:
+                translated_coeffs = [
                     sum(derivatives[m + j + self.order + src_expansion.order]
-                        * src_coeff_exprs[src_expansion.get_storage_index(m)]
-                        for m in src_expansion.get_coefficient_identifiers()))
+                            * src_coeff_exprs[src_expansion.get_storage_index(m)]
+                        for m in src_expansion.get_coefficient_identifiers())
+                    for j in self.get_coefficient_identifiers()]
+
+            if m2l_translation_classes_dependent_data is None:
+                translated_coeffs = self.m2l_postprocess_local_exprs(src_expansion,
+                        translated_coeffs, src_rscale, tgt_rscale, sac)
 
-            translated_coeffs = self.m2l_postprocess_local_exprs(src_expansion,
-                translated_coeffs, src_rscale, tgt_rscale, sac)
             return translated_coeffs
 
         raise RuntimeError("do not know how to translate %s to %s"
                            % (type(src_expansion).__name__,
                                type(self).__name__))
 
+    def loopy_translate_from(self, src_expansion):
+        if isinstance(src_expansion, self.mpole_expn_class):
+            ncoeff_src = self.m2l_preprocess_multipole_nexprs(src_expansion)
+            ncoeff_tgt = self.m2l_postprocess_local_nexprs(src_expansion)
+
+            icoeff_src = pymbolic.var("icoeff_src")
+            icoeff_tgt = pymbolic.var("icoeff_tgt")
+            domains = [f"{{[icoeff_tgt]: 0<=icoeff_tgt<{ncoeff_tgt} }}"]
+
+            src_coeffs = pymbolic.var("src_coeffs")
+            m2l_translation_classes_dependent_data = pymbolic.var("data")
+
+            if self.use_preprocessing_for_m2l:
+                expr = src_coeffs[icoeff_tgt] \
+                        * m2l_translation_classes_dependent_data[icoeff_tgt]
+            else:
+                expr = src_coeffs[icoeff_src] \
+                       * m2l_translation_classes_dependent_data[
+                               icoeff_tgt + icoeff_src]
+                domains += [
+                        f"{{[icoeff_src]: 0<=icoeff_src<{ncoeff_src} }}"]
+
+            insns = f"""
+            coeff[icoeff_tgt] = coeff[icoeff_tgt] + {expr}
+            """
```

Comment:
As above.

Reply:
Done in 1e3f1e5c494121f704ce34ffde21c395704bd424